### PR TITLE
[Merged by Bors] - feat(NumberTheory/Ostrowski): Ostrowski's Theorem for the rationals

### PR DESCRIPTION
**Pull request by fbarroero**
We complete the proof of the Archimedean case of Ostrowski's Theorem for the rational numbers:
```lean
theorem mulRingNorm_equiv_standard_of_unbounded (notbdd: ¬ ∀ (n : ℕ), f n ≤ 1) :
    MulRingNorm.equiv f mulRingNorm_real
```

This concludes the proof of the theorem:
```lean
theorem mulRingNorm_equiv_standard_or_padic (f : MulRingNorm ℚ) (hf_nontriv : f ≠ 1) :
    (MulRingNorm.equiv f mulRingNorm_real) ∨
    ∃! p, ∃ (hp : Fact (p.Prime)), MulRingNorm.equiv f (mulRingNorm_padic p)
```

Co-authored-by: David Kurniadi Angdinata [dka31@cantab.ac.uk](mailto:dka31@cantab.ac.uk)
Co-authored-by: Laura Capuano [laura.capuano@uniroma3.it](mailto:laura.capuano@uniroma3.it)
Co-authored-by: Nirvana Coppola [nirvanac93@gmail.com](mailto:nirvanac93@gmail.com)
Co-authored-by: María Inés de Frutos Fernández [maria.defrutos@uam.es](mailto:maria.defrutos@uam.es)
Co-authored-by: Sam van Gool [vangool@irif.fr](mailto:vangool@irif.fr)
Co-authored-by: Silvain Rideau-Kikuchi [silvain.rideau-kikuchi@ens.fr](mailto:silvain.rideau-kikuchi@ens.fr)
Co-authored-by: Amos Turchet [amos.turchet@uniroma3.it](mailto:amos.turchet@uniroma3.it)
Co-authored-by: Francesco Veneziano [veneziano@dima.unige.it](mailto:veneziano@dima.unige.it)

---

- [x] depends on: #14442

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by riccardobrasca** [comment]
Can you please fix the conflict? Thanks!

**Comment by fbarroero** [comment]
> Can you please fix the conflict? Thanks!

Done!

File: Mathlib/NumberTheory/Ostrowski.lean [hunk 1]
```diff
@@ -261,6 +285,23 @@ section Archimedean
...
+  neg'      := norm_neg
+  eq_zero_of_map_eq_zero' := by simp only [abs_eq_zero, Rat.cast_eq_zero, imp_self, forall_const]
+  map_one' := by simp only [Rat.cast_one, abs_one]
+  map_mul' := by
+    simp only [Rat.cast_mul]
+    exact_mod_cast abs_mul
+}
+
+@[simp] lemma mul_ring_norm_eq_abs (r : ℚ) : mulRingNorm_real r = |r| := by
+  simp only [Rat.cast_abs]
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/Ostrowski.lean:302]: Is this line necessary?

File: Mathlib/NumberTheory/Ostrowski.lean [hunk 2]
```diff
@@ -343,6 +384,133 @@ lemma one_lt_of_not_bounded (notbdd : ¬ ∀ n : ℕ, f n ≤ 1) {n₀ : ℕ} (h
...
+variable {m n : ℕ} (hm : 1 < m) (hn : 1 < n) (notbdd : ¬ ∀ (n : ℕ), f n ≤ 1)
+
+include hm notbdd in
+private lemma expr_pos : 0 < m * f m / (f m - 1) := by
+  apply div_pos (mul_pos (mod_cast zero_lt_of_lt hm)
+      (map_pos_of_ne_zero f (mod_cast ne_zero_of_lt hm)))
+  linarith only [one_lt_of_not_bounded notbdd hm]
+
+include hn hm notbdd in
+private lemma param_upperbound (k : ℕ) (hk : k ≠ 0) :
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/Ostrowski.lean:398]:
```suggestion
private lemma param_upperbound {k : ℕ} (hk : k ≠ 0) :
```

File: Mathlib/NumberTheory/Ostrowski.lean [hunk 3]
```diff
@@ -343,6 +384,133 @@ lemma one_lt_of_not_bounded (notbdd : ¬ ∀ n : ℕ, f n ≤ 1) {n₀ : ℕ} (h
...
+      (mod_cast Nat.ne_of_lt' oneltm) (by linarith only [hm])]
+    have hfn : f n = n ^ (Real.logb n (f n)) := by
+      rw [Real.rpow_logb (mod_cast zero_lt_of_lt h1) (mod_cast Nat.ne_of_lt' h1)
+      (by apply map_pos_of_ne_zero; exact_mod_cast not_eq_zero_of_lt h1)]
+    rwa [← hs, symmetric_roles oneltm h1 notbdd hfm hfn]
+
 end Archimedean
 
+/-- Ostrowski's Theorem -/
+theorem mulRingNorm_equiv_standard_or_padic (f : MulRingNorm ℚ) (hf_nontriv : f ≠ 1) :
```
Comment (riccardobrasca) [inline Mathlib/NumberTheory/Ostrowski.lean:507]: Can you add the theorem to the doc at the beginning of the file?

**Review by riccardobrasca — APPROVED** [review]

File: Mathlib/NumberTheory/Ostrowski.lean (hunk 1 above)
Comment (fbarroero) [reply Mathlib/NumberTheory/Ostrowski.lean:302]: It is used at some point in `mulRingNorm_equiv_standard_of_unbounded`

File: Mathlib/NumberTheory/Ostrowski.lean (hunk 3 above)
Comment (fbarroero) [reply Mathlib/NumberTheory/Ostrowski.lean:513]: Done!

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**